### PR TITLE
Fixing branch

### PR DESCRIPTION
**Pull request by busterbeam**
Good work by @EarthKiii should not go unrewarded

### CODE REVIEW

File: jigsaw/__main__.py [hunk 1]
```diff
@@ -7,6 +7,7 @@
 from random import choice
 
 from client import GameClient
+from client.menu import Menu
```
Comment (LukeHankey) [inline jigsaw/__main__.py:9]:
```suggestion
from .menu import Menu
```

File: jigsaw/client/credits_view.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
+from pygame import font, Rect, quit
+import pygame_gui
+from client.constants import screen, WHITE, BLUE, WIDTH, manager, HEIGHT
```
Comment (LukeHankey) [inline jigsaw/client/credits_view.py:3]:
```suggestion
from .constants import screen, WHITE, BLUE, WIDTH, manager, HEIGHT
```

File: jigsaw/client/credits_view.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    def __init__(self):
+        self.title_font = font.Font(None, 48)
+
+        self.title = self.title_font.render("Jigsaw Game", True, BLUE)
+        self.contributors = [
+            "@LukeHankey LukeHankey Luke",
+            "@busterbeam busterbeam Nathan Lloyd",
+            "@EarthKiii EarthKiii Jonas Charrier",
+            "@DavidStrootman DavidStrootman David Strootman",
+            "@bartoszkobylinski Bartosz Kobylinski",
```
Comment (LukeHankey) [inline jigsaw/client/credits_view.py:17]:
```suggestion
            "@bartoszkobylinski Bartosz Kobylinski",
            "@chuchumuru Pilogic
```

File: jigsaw/client/credits_view.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
+from pygame import font, Rect, quit
+import pygame_gui
+from client.constants import screen, WHITE, BLUE, WIDTH, manager, HEIGHT
+from pygame.locals import QUIT, USEREVENT
+
+
+class CreditsView:
+    def __init__(self):
```
Comment (LukeHankey) [inline jigsaw/client/credits_view.py:8]:
```suggestion
    def __init__(self) -> None:
```

File: jigsaw/client/credits_view.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    def event_handler(self, e):
+        if e.type == QUIT:
+            quit()
+        if e.type == USEREVENT:
+            if e.user_type == pygame_gui.UI_BUTTON_PRESSED:
+                if e.ui_element == self.back_button:
+                    return False
+        return True
+
+    def show(self):
```
Comment (LukeHankey) [inline jigsaw/client/credits_view.py:52]:
```suggestion
    def show(self) -> None:
```

File: jigsaw/client/credits_view.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        self.credit_text_box.show()
+        self.back_button.show()
+
+    def hide(self):
```
Comment (LukeHankey) [inline jigsaw/client/credits_view.py:56]:
```suggestion
    def hide(self) -> None:
```

File: jigsaw/client/credits_view.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
+from pygame import font, Rect, quit
+import pygame_gui
+from client.constants import screen, WHITE, BLUE, WIDTH, manager, HEIGHT
+from pygame.locals import QUIT, USEREVENT
```
Comment (LukeHankey) [inline jigsaw/client/credits_view.py:4]:
```suggestion
from pygame.locals import QUIT, USEREVENT
from pygame.event import Event
```

File: jigsaw/client/credits_view.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+        self.back_button = pygame_gui.elements.UIButton(
+            relative_rect=Rect((WIDTH // 2 - 50, HEIGHT // 2 + 60), (100, 50)),
+            text="Back",
+            manager=manager,
+        )
+
+        self.hide()
+
+    def event_handler(self, e):
```
Comment (LukeHankey) [inline jigsaw/client/credits_view.py:43]:
```suggestion
    def event_handler(self, e: Event) -> bool:
```

File: jigsaw/client/menu.py [hunk 9]
```diff
@@ -0,0 +1,167 @@
+import random
+
+import pygame_gui
+from pygame import init, draw, Rect, display, time, event, quit
+from pygame.locals import QUIT, USEREVENT
+
+from client.constants import WHITE, BLUE, WIDTH, HEIGHT, screen, font, manager
+from client.credits_view import CreditsView
+from client.settings_view import SettingsView
```
Comment (LukeHankey) [inline jigsaw/client/menu.py]:
```suggestion
from .constants import WHITE, BLUE, WIDTH, HEIGHT, screen, font, manager
from .credits_view import CreditsView
from .settings_view import SettingsView
```

File: jigsaw/client/menu.py [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+
+# Screen dimensions
+WIDTH, HEIGHT = 800, 600
+screen = pygame.display.set_mode((WIDTH, HEIGHT))
+font = pygame.font.SysFont(None, 36)  # Default font, size 36
+overlap_sensitivity = 5
+
+pygame.display.set_caption('Game Menu')
+
+'''
```
Comment (LukeHankey) [inline jigsaw/client/menu.py]: Is this needed?

File: jigsaw/client/menu.py [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+
+
+class PlayGameView:
+    pass
+
+class Menu:
+    def __init__(self):
```
Comment (LukeHankey) [inline jigsaw/client/menu.py]:
```suggestion
    def __init__(self) -> None:
```

File: jigsaw/client/menu.py [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+            button = pygame_gui.elements.UIButton(
+                relative_rect=rectangle, text=mode.title(), manager=manager)
+            count += 83
+            setattr(self, f"{mode}_button", button)
+
+        self.right_panel = pygame_gui.elements.UIPanel(
+            relative_rect=Rect((WIDTH - 200, 0), (200, HEIGHT)),
+            starting_height=0,
+            manager=manager
+        )
```
Comment (LukeHankey) [inline jigsaw/client/menu.py:56]:
```suggestion
    def build(self) -> None:
        button_size = (100, 50)
        right_edge = WIDTH - button_size[0]
        
        count = 83
        MODES = ["quick", "secret", "settings", "credits"]
        
        for mode in MODES:
            rectangle = Rect((right_edge - 50, count), button_size)
            button = pygame_gui.elements.UIButton(
                relative_rect=rectangle, text=mode.title(), manager=manager
            )
            count += 83
            setattr(self, f"{mode}_button", button)

        self.right_panel = pygame_gui.elements.UIPanel(
            relative_rect=Rect((WIDTH - 200, 0), (200, HEIGHT)),
            starting_height=0,
            manager=manager
        )
```

File: jigsaw/client/menu.py [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+
+    def draw_completion_bar(self, completion_percent, position, size):
+        """ Draws a completion bar on the screen.
+        Args:
```
Comment (LukeHankey) [inline jigsaw/client/menu.py:60]:
```suggestion
    def draw_completion_bar(self, completion_percent: float, position: tuple[int, int], size: tuple[int, int]) -> None:
        """ Draws a completion bar on the screen.
        
        Args:
```

File: jigsaw/client/menu.py [hunk 14]
```diff
@@ -0,0 +1,167 @@
+import random
+
+import pygame_gui
+from pygame import init, draw, Rect, display, time, event, quit
+from pygame.locals import QUIT, USEREVENT
```
Comment (LukeHankey) [inline jigsaw/client/menu.py:5]:
```suggestion
from pygame.locals import QUIT, USEREVENT
from pygame.event import Event
```

File: jigsaw/client/menu.py [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+        """
+        border_color = (0, 0, 0)
+        fill_color = (0, 255, 0)
+
+        draw.rect(screen, border_color, position + size, 2)
+
+        fill_width = int(size[0] * completion_percent)
+        draw.rect(screen, fill_color, (position[0], position[1], fill_width, size[1]))
+
+    def event_handler(self, e):
```
Comment (LukeHankey) [inline jigsaw/client/menu.py:74]:
```suggestion
    def event_handler(self, e: Event) -> Literal[True]:
```

File: jigsaw/client/menu.py [hunk 16]
```diff
@@ -0,0 +1,167 @@
+import random
+
```
Comment (LukeHankey) [inline jigsaw/client/menu.py:2]:
```suggestion

from typing import Literal, NoReturn
```

File: jigsaw/client/menu.py [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+                    self.active_view = self.view_instances["play_game"]
+                if e.ui_element == self.settings_button:
+                    self.active_view = self.view_instances["settings"]
+                    self.active_view.show()
+                if e.ui_element == self.credits_button:
+                    self.active_view = self.view_instances["credits"]
+                    self.active_view.show()
+        return True
+
+    def mainloop(self):
```
Comment (LukeHankey) [inline jigsaw/client/menu.py:92]:
```suggestion
    def mainloop(self) -> NoReturn:
```

File: jigsaw/client/menu.py [hunk 18]
```diff
@@ -0,0 +1,167 @@
...
+                    self.active_view.hide()
+                    self.active_view = self
+                manager.process_events(e)
+            manager.draw_ui(screen)
+            manager.update(time_delta)
+            display.update()
+            display.flip()
+
+
+def hamming_distance(current_positions, correct_positions):
```
Comment (LukeHankey) [inline jigsaw/client/menu.py:107]:
```suggestion
def hamming_distance(current_positions: list[int], correct_positions: list[int]) -> int:
```

File: jigsaw/client/menu.py [hunk 19]
```diff
@@ -0,0 +1,167 @@
...
+    """Calculate the Hamming distance between the current puzzle positions and the correct positions."""
+    return sum(current != correct for current, correct in zip(current_positions, correct_positions))
+
+
+def simulate_tile_positions(total_tiles):
```
Comment (LukeHankey) [inline jigsaw/client/menu.py:112]:
```suggestion
def simulate_tile_positions(total_tiles: int) -> list[int]:
```

File: jigsaw/client/menu.py [hunk 20]
```diff
@@ -0,0 +1,167 @@
...
+    """
+    Simulate random tile position to produce random Hamming distance
+    """
+    return random.sample(range(total_tiles), total_tiles)
+
+
+# Views
+
+def play_game(difficulty):
```
Comment (LukeHankey) [inline jigsaw/client/menu.py]:
```suggestion
def play_game(difficulty) -> dict[str, str]:
```
Please add what the type of `difficulty` is. I can't see where this function is used currently so not sure on the type.

File: jigsaw/client/settings_view.py [hunk 21]
```diff
@@ -0,0 +1,96 @@
...
+import tkinter as tk
+from tkinter import filedialog
+
+from pygame.locals import USEREVENT
+
+from client.constants import screen, WHITE, BLUE, font, WIDTH, HEIGHT, manager
+
+
+class SettingsView:
+    def __init__(self):
```
Comment (LukeHankey) [inline jigsaw/client/settings_view.py:15]:
```suggestion
    def __init__(self) -> None:
```

File: jigsaw/client/settings_view.py [hunk 22]
```diff
@@ -0,0 +1,96 @@
...
+        self.settings = {"fullscreen": False, "overlap": 0.25, "image": ""}
+        self.title = font.render("Settings", True, BLUE)
+        self.build()
+
+    def build(self):
```
Comment (LukeHankey) [inline jigsaw/client/settings_view.py:20]:
```suggestion
    def build(self) -> None:
```

File: jigsaw/client/settings_view.py [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
+
+                if e.ui_element == self.back_button:
+                    self.hide()
+                    return False
+        return True
+
+    def show(self):
+        self.settings_panel.show()
+
+    def hide(self):
```
Comment (LukeHankey) [inline jigsaw/client/settings_view.py:95]:
```suggestion
    def hide(self) -> None:
```

File: jigsaw/client/settings_view.py [hunk 24]
```diff
@@ -0,0 +1,96 @@
...
+                    self.open_file_dialog()
+                    if self.settings["image"]:
+                        print(f'File selected: {self.settings["image"]}')
+
+                if e.ui_element == self.back_button:
+                    self.hide()
+                    return False
+        return True
+
+    def show(self):
```
Comment (LukeHankey) [inline jigsaw/client/settings_view.py:92]:
```suggestion
    def show(self) -> None:
```

File: jigsaw/client/settings_view.py [hunk 25]
```diff
@@ -0,0 +1,96 @@
...
+
+    def open_file_dialog(self):
+        if sys.platform in ['linux', 'darwin']:
+            self.selected_file = os.popen('zenity --file-selection').read().strip()
+        elif sys.platform == 'win32':
+            root = tk.Tk()
+            root.withdraw()
+            self.selected_file = filedialog.askopenfilename()
+
+    def event_handler(self, e):
```
Comment (LukeHankey) [inline jigsaw/client/settings_view.py:74]:
```suggestion
    def event_handler(self, e: Event) -> bool:
```

File: jigsaw/client/settings_view.py [hunk 26]
```diff
@@ -0,0 +1,96 @@
...
+            relative_rect=Rect((50, 370), (200, 50)),
+            value_range=(0.15, 0.25),
+            start_value=0.25,
+            manager=manager,
+            container=self.settings_panel
+        )
+
+        self.settings_panel.hide()
+
+    def open_file_dialog(self):
```
Comment (LukeHankey) [inline jigsaw/client/settings_view.py:66]:
```suggestion
    def open_file_dialog(self) -> None:
```

File: jigsaw/client/settings_view.py [hunk 27]
```diff
@@ -0,0 +1,96 @@
+import os
+import sys
+from pygame import Rect, quit
+import pygame_gui
+from pygame import QUIT
+import tkinter as tk
+from tkinter import filedialog
+
+from pygame.locals import USEREVENT
```
Comment (LukeHankey) [inline jigsaw/client/settings_view.py:9]:
```suggestion
from pygame.locals import USEREVENT
from pygame.event import Event
```

**Review by LukeHankey — CHANGES_REQUESTED** [review]

File: jigsaw/__main__.py [hunk 28]
```diff
@@ -6,44 +6,21 @@
...
-    "solved_tiles": correct_tiles,
-    "scrambled_tiles": tile_scrambler(correct_tiles),
-}
+menu = Menu(filename)
+menu.mainloop()
 
 with open(LOG_FILENAME, "w") as log_file:
     dump(game_log, log_file, indent=1)
 
-game = GameClient(action)
```
Comment (busterbeam) [inline jigsaw/__main__.py]: are we putting everything into Menu?

I said from the very beinging launcher window, which then moves to Gameclient window, pass options along

File: jigsaw/client/credits_view.py (hunk 2 above)
Comment (busterbeam) [reply jigsaw/client/credits_view.py:3]: colors can be strings, drop those dumb color constants, plus the manager is an actual runtime object which shouldn't be passed around like that

File: jigsaw/client/menu.py (hunk 13 above)
Comment (busterbeam) [reply jigsaw/client/menu.py:60]: this is a game client thing, for when someones solving the puzzle move it over there

File: jigsaw/client/menu.py (hunk 18 above)
Comment (busterbeam) [reply jigsaw/client/menu.py:107]: duplicate function, already in utils.py

File: jigsaw/client/menu.py (hunk 19 above)
Comment (busterbeam) [reply jigsaw/client/menu.py:112]: just delete

**Review by busterbeam — COMMENTED** [review]
all good stuff

**Review by LukeHankey — APPROVED** [review]

**Pull request merged**